### PR TITLE
NOT MERGEABLE: graceful cleanup on Application dtor for NodeList

### PR DESCRIPTION
**Pull request by birarda**

### CODE REVIEW

**Comment by birarda** [comment]
Code here is ready to go, but need some input from @Atlante45.

Seems like DependencyManager is still trying to call the destructor on LimitedNodeList even though the NodeList has been manually deleted by calling deleteLater(). Is there a solution to this? 

I could make the LimitedNodeList custom deleter be `deleteLater` like for the AudioClient and then use DependencyManager::destroy<NodeList> from Application, but in the DomainServer we use the LimitedNodeList on the main thread, and the custom deleter will be called once the application event loop is gone. At that point I'm not sure if the destructor for LimitedNodeList would ever get called (I'm going to test if Qt is smart enough to delete immediately in this case).

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5d9a3811ca2b06d0baf0222875d4f7f3b0f9710f available at: https://gist.github.com/fef7dbe97dc907ff8043<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7a8e94f1e572fc927b4c3ecc6b5453384ddc1572 available at: https://gist.github.com/fef7dbe97dc907ff8043<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.dmg

**Comment by birarda** [comment]
I think this is a DependencyManager bug @Atlante45 - even if I make the LimitedNodeList deleter `deleteLater`, if I call `DependencyManager::destroy<NodeList>` it looks like the DependencyManager thinks it still has a LimitedNodeList to destroy.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -612,6 +600,14 @@ Application::~Application() {
     DependencyManager::destroy<GeometryCache>();
     //DependencyManager::destroy<ScriptCache>();
     DependencyManager::destroy<SoundCache>();
+    
+    auto nodeList = DependencyManager::get<NodeList>();
+    QThread* nodeThread = nodeList->thread();
+    nodeList->deleteLater();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: the problem is that DependencyManager still has NodeList in it's hash... it has no way of knowing you've deleted it. Seems like you could add something like:
```
DependencyManager::destroyLater<T>();
```
That removes from the hash and calls deleteLater()

**Comment by birarda** [comment]
Yeah, or a way to tell the DependencyManager - hey, I've deleted this thing manually. 

If you just set it to NULL it tries to delete what it thought it had before.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -612,6 +600,14 @@ Application::~Application() {
...
+    
+    // ask the node thread to quit and wait until it is done
+    nodeThread->quit();
+    nodeThread->wait();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: when is nodeThread being deleted? I don't see you calling delete, deleteLater or hooking it up to a QObject as a parent.

Comment (birarda) [inline interface/src/Application.cpp]: whoops - missed that, thanks

File: interface/src/Application.cpp (hunk 1 above)
Comment (birarda) [inline interface/src/Application.cpp]: Actually the DM uses a weak pointer and does know when you've deleted something. The issue is that if I manually delete the NodeList then it doesn't realize it doesn't actually have a LimitedNodeList and in trying to delete the LNL it produces a crash - but it does realize the NodeList is gone.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d7e99594aeaab9d984b5f13f27da7abbbd031e8e available at: https://gist.github.com/fef7dbe97dc907ff8043<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4429.dmg

File: interface/src/Application.cpp (hunk 1 above)
Comment (ZappoMan) [inline interface/src/Application.cpp]: I think we're saying the same thing here. DependencyManager needs a way of having LNL removed from it's hash.

Comment (birarda) [inline interface/src/Application.cpp]: "the problem is that DependencyManager still has NodeList in it's hash"

The crash is not from the NL still being in its hash, it's from the LNL still being in the hash.

**Pull request merged**